### PR TITLE
[OTEL-2897][OTAGENT-775] Add feature gate to avoid duplicating container tags in OTLP traces

### PR DESCRIPTION
**Pull request by jade-guiton-dd**
### What does this PR do?

OTel trace pipelines in the Agent (either OTLP ingest or DDOT) use the Agent's tagger component to enrich the traces with container tags. At the moment, this is done in two places:
- Calling the tagger is the infraattributes processor's purpose; it is the only way container tags are extracted for metrics/logs. It also has more features (tags for non-container entities, container ID inference). The extracted tags are emitted as simple resource attributes, which get mapped to span attributes.
- The trace agent's OTLP receiver also calls the tagger. It is more limited, but it correctly outputs the extracted tags as container tags (under `_dd.tags.container`).

Both have been enabled in DDOT since its inception, and in the OTLP ingest pipeline since #40203. This leads to redundant work being done, and to increased traffic volume, as container tags are duplicated across span attributes / container tags. (In some cases even triplicated: `container.id` becomes a duplicated `container_id` but is never removed).

This PR adds a feature flag `enable_otlp_container_tags_v2` which:
- disables the call to the tagger in the trace agent;
- promotes the Datadog-convention resource attributes emitted by the processor to container tags;
- filters out resource attributes which get promoted into container tags so they don't get added as span attributes as well.

**Update**: We decided not to filter out OTel-convention resource attributes, as OTel customers likely want to be able to search for them in the application. Filtering out Datadog-convention tags from the Infra Attributes Processor should be enough to solve the duplication issue.

While it's unrelated to container tag duplication, I also took the opportunity to remove the code that sourced a container ID from the Kubernetes Pod UID attribute, which AFAIK was a historical mishap.

[Link to an Excalidraw diagram illustrating the changes](https://excalidraw.com/#json=jVshUOwsgsQppmlXOS2i9,oEdIFb60jXBR6M0csvbq8g)

### Motivation

This is meant as a long-term fix for [OTAGENT-689](https://datadoghq.atlassian.net/browse/OTAGENT-689).

### Describe how you validated your changes

I added a test validating that, with the feature gate enabled and an infraattributes processor upstream, we see the expected container tags, but no copies under span attributes.

Because that test is essentially an integration test, it has fairly widely-ranging dependencies, so I added it into a new `pkg/trace/otel/integration` package to avoid too many transitive dependencies in `pkg/trace`. I can move it if there is a better place for it.

@RiantZ also performed some manual testing and verified that the feature flag behaves as expected when enabled (removing span attributes that are duplicates of container tags, lowering output traffic).

### Additional Notes

[OTAGENT-689]: https://datadoghq.atlassian.net/browse/OTAGENT-689?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pkg/trace/api/otlp.go [hunk 1]
```diff
@@ -273,45 +273,6 @@ func (o *OTLPReceiver) receiveResourceSpansV2(ctx context.Context, rspans ptrace
...
-	}
-
-	tags := tagstats.AsTags()
-	_ = o.statsd.Count("datadog.trace_agent.otlp.spans", spancount, tags, 1)
-	_ = o.statsd.Count("datadog.trace_agent.otlp.traces", int64(len(tracesByID)), tags, 1)
-	p := Payload{
-		Source:                 tagstats,
-		ClientComputedStats:    traceutil.GetOTelAttrVal(resourceAttributes, true, keyStatsComputed) != "" || clientComputedStats,
-		ClientComputedTopLevel: o.conf.HasFeature("enable_otlp_compute_top_level_by_span_kind"),
-	}
```
Comment (jade-guiton-dd) [inline pkg/trace/api/otlp.go:317]: The diff may look big, but I moved this code almost as-is to the end of the function, so that the resource attributes removed by `RemoveContainerTagsFromResourceAttributes` don't show up in the span attributes set in `OtelSpanToDDSpan`.

As far as I can tell, this computation of `tracesByID`/`lang` is completely independent of `hostname`/`containerID`/`env`/`containerTags`, so the order shouldn't affect the behavior when the feature gate is off.

File: pkg/trace/api/otlp_test.go [hunk 2]
```diff
@@ -4615,3 +4622,78 @@ func TestConvertSpanDBNameMapping(t *testing.T) {
...
+	for _, rs := range td.ResourceSpans().All() {
+		_, err := o.rcv.ReceiveResourceSpans(ctx, rs, http.Header{}, nil)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func TestContainerTagsV2(t *testing.T) {
```
Comment (jade-guiton-dd) [inline pkg/trace/api/otlp_test.go]: This is essentially an integration test between the `infraattributesprocessor` and the trace agent. There may be a better place to put this that doesn't add as many dependencies to `pkg/trace`.

File: releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    Added ``enable_otlp_container_tags_v2`` feature gate, which reduces duplication of container tags in OTLP traces.
```
Comment (domalessi) [inline releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml]:
```suggestion
    Added the `enable_otlp_container_tags_v2` feature gate, which reduces duplication of container tags in OTLP traces.
```

**Review by domalessi — APPROVED** [review]

File: pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go [hunk 4]
```diff
@@ -296,6 +304,78 @@ func ContainerTagsFromResourceAttributes(attrs pcommon.Map) map[string]string {
...
+func RemoveContainerTagsFromResource(res pcommon.Resource) (map[string]string, pcommon.Resource) {
+	ddtags := make(map[string]string)
+	tagSources := make(map[string]containerTagSource)
+	filteredRes := pcommon.NewResource()
+	res.CopyTo(filteredRes)
+
+	filteredRes.Attributes().RemoveIf(func(key string, value pcommon.Value) bool {
+		valueStr := value.Str()
+		if valueStr == "" {
+			return false
```
Comment (dineshg13) [inline pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go]: I see need for this only in OTLP . For DDOT, since it initializes a different trace Agent component, we can remove tagger there.

File: releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml (hunk 3 above)
Comment (dineshg13) [reply releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml]: This isn't feature gate rt ? It is just a another feature flag of Agent ??

File: pkg/trace/api/otlp.go [hunk 5]
```diff
@@ -340,45 +301,97 @@ func (o *OTLPReceiver) receiveResourceSpansV2(ctx context.Context, rspans ptrace
 
 	containerID := traceutilotel.GetOTelAttrVal(resourceAttributes, true, transform.KeyDatadogContainerID)
 	if containerID == "" && !o.conf.OTLPReceiver.IgnoreMissingDatadogFields {
-		containerID = traceutilotel.GetOTelAttrVal(resourceAttributes, true, string(semconv.ContainerIDKey), string(semconv.K8SPodUIDKey))
+		if o.conf.HasFeature("enable_otlp_container_tags_v2") {
+			containerID = traceutilotel.GetOTelAttrVal(resourceAttributes, true, string(semconv.ContainerIDKey))
+		} else {
+			containerID = traceutilotel.GetOTelAttrVal(resourceAttributes, true, string(semconv.ContainerIDKey), string(semconv.K8SPodUIDKey))
```
Comment (dineshg13) [inline pkg/trace/api/otlp.go]: Why the default is to duplicate the tags ?

**Review by dineshg13 — CHANGES_REQUESTED** [review]

File: pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go (hunk 4 above)
Comment (jade-guiton-dd) [reply pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go]: The call to the tagger is `getContainerTags(o.conf.ContainerTags, containerID)` in `otlp.go`, not in this function. It is disabled for both OTLP ingest and DDOT when the feature flag is enabled.

Could you elaborate on "a different trace Agent component"? As far as I know, both the trace agent process used by OTLP ingest and the instance used by DDOT's exporter go through `receiveResourceSpansV2`

File: releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml (hunk 3 above)
Comment (jade-guiton-dd) [reply releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml]: Right sorry, I got the terminology wrong. I'll update that as well.

(The double backquotes are intentional however, I got [this CI failure](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/1378131165) when only using single quotes)

File: pkg/trace/api/otlp.go (hunk 5 above)
Comment (jade-guiton-dd) [reply pkg/trace/api/otlp.go]: This particular part of the code has nothing to do with tag duplication. It just determines which resource attributes we search when looking for the container ID. Historically, our code will use `k8s.pod.uid` as a fallback for `container.id`, which AFAIK makes no sense and is a historical accident. So since I was modifying container tag behavior in breaking ways already, I took the opportunity to remove that fallback.

Comment (jade-guiton-dd) [reply pkg/trace/api/otlp.go]: As for why the new behavior is under an `enable` feature flag: the change to container tag resolution is pretty sensitive and I'm not 100% certain of the consequences, so I figured it would be better to start with that, check with the relevant customers if it solves their problem and there are no adverse effects, then turn it on by default

**Review by RiantZ — APPROVED** [review]

File: pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go [hunk 6]
```diff
@@ -296,6 +304,78 @@ func ContainerTagsFromResourceAttributes(attrs pcommon.Map) map[string]string {
...
+// 3. Datadog semantic conventions (pre-mapped tags, usually from the infraattributes processor).
+//
+// Only string-type resource attributes will be extracted as container tags.
+// In the case of duplicates between the three sources, OTel conventions take priority over custom tags,
+// which take priority over pre-mapped tags.
+//
+// This function also returns a copy of the input resource, with sources 2 and 3 filtered out.
+// This filtered resource should be used when converting spans to Datadog format,
+// to avoid attributes being duplicated as both span attributes and container tags.
+func RemoveContainerTagsFromResource(res pcommon.Resource) (map[string]string, pcommon.Resource) {
```
Comment (nathan-b) [inline pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go]: Recommend writing a unit test for this function. A unit test provides better coverage of edge cases and documents the intended behavior better than the integration test.

File: pkg/trace/api/otlp.go [hunk 7]
```diff
@@ -340,45 +301,97 @@ func (o *OTLPReceiver) receiveResourceSpansV2(ctx context.Context, rspans ptrace
 
 	containerID := traceutilotel.GetOTelAttrVal(resourceAttributes, true, transform.KeyDatadogContainerID)
 	if containerID == "" && !o.conf.OTLPReceiver.IgnoreMissingDatadogFields {
-		containerID = traceutilotel.GetOTelAttrVal(resourceAttributes, true, string(semconv.ContainerIDKey), string(semconv.K8SPodUIDKey))
+		if o.conf.HasFeature("enable_otlp_container_tags_v2") {
```
Comment (nathan-b) [inline pkg/trace/api/otlp.go:303]: You might need to apply this logic to https://github.com/DataDog/datadog-agent/blob/758f0dc3a0af66da9a34d49558e4ddcd4b5dea14/pkg/trace/transform/transform.go#L302 as well. WDYT?

File: pkg/trace/api/otlp.go [hunk 8]
```diff
@@ -340,45 +301,97 @@ func (o *OTLPReceiver) receiveResourceSpansV2(ctx context.Context, rspans ptrace
...
-				// but we have some information from the source provider that we can add
-				appendTags(payloadTags, src.Tag())
+			builder = flatten(attributes.ContainerTagsFromResourceAttributes(resourceAttributes))
+
+			// Populate container tags by calling ContainerTags tagger from configuration
+			if tags := getContainerTags(o.conf.ContainerTags, containerID); tags != "" {
+				appendTags(builder, tags)
+			} else {
+				// we couldn't obtain any container tags
+				if src.Kind == source.AWSECSFargateKind {
```
Comment (nathan-b) [inline pkg/trace/api/otlp.go:337]: The v2 path lacks the Fargate fallback. Is this intentional? I would expect container tags v2 to apply to Fargate tasks as well?

File: pkg/trace/otel/integration/container_tags_v2_test.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	for _, rs := range td.ResourceSpans().All() {
+		_, err := o.rcv.ReceiveResourceSpans(ctx, rs, http.Header{}, nil)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func TestContainerTagsV2(t *testing.T) {
```
Comment (nathan-b) [inline pkg/trace/otel/integration/container_tags_v2_test.go:56]: This test doesn't seem to cover the filtering behavior (verifying OTel convention attrs stay but pre-mapped ones don't).

File: pkg/trace/api/otlp.go (hunk 7 above)
Comment (jade-guiton-dd) [reply pkg/trace/api/otlp.go:303]: I agree that we should remove the fallback to `k8s.pod.uid` in there as well at some point. It seems more difficult though, since that function doesn't have access to the Agent feature flags. I assume that for the time being it's probably fine if the stat payloads keep incorrect container IDs while the corresponding spans don't?

File: pkg/trace/api/otlp.go (hunk 8 above)
Comment (jade-guiton-dd) [reply pkg/trace/api/otlp.go:337]: Yes, this was intentional. As far as I can tell, that logic is redundant with the general mapping from `aws.ecs.task.arn` to `task_arn`, so I'm not sure if it changes any behavior?

File: pkg/trace/otel/integration/container_tags_v2_test.go (hunk 9 above)
Comment (jade-guiton-dd) [reply pkg/trace/otel/integration/container_tags_v2_test.go:56]: What do you mean? The test explicitly checks that `container.id` is passed through but `container_id` is filtered out.

File: pkg/trace/api/otlp.go (hunk 8 above)
Comment (RiantZ) [reply pkg/trace/api/otlp.go:337]: Well, if I understood correctly @jade-guiton-dd wasn't modifying this logic, it was just moved from one place to another, to be capable to handle filtered tags properly.
One of the prev. PR which was working with the logic is: https://github.com/DataDog/datadog-agent/pull/33753/changes

File: pkg/trace/api/otlp.go (hunk 7 above)
Comment (nathan-b) [reply pkg/trace/api/otlp.go:303]: I don't know the answer to that question definitively. Intuitively I would expect it to be a problem when different payloads have different container IDs for the same logical container, but my intuition could be wrong here.

File: pkg/trace/api/otlp.go (hunk 8 above)
Comment (nathan-b) [reply pkg/trace/api/otlp.go:337]: Fair enough, I didn't look into what the Fargate fallback actually returned and whether those data were still necessary!

File: pkg/trace/otel/integration/container_tags_v2_test.go (hunk 9 above)
Comment (nathan-b) [reply pkg/trace/otel/integration/container_tags_v2_test.go:56]: Ah, you're right, I mis-read the test. Sorry for the noise on this one!

File: comp/otelcol/otlp/components/processor/infraattributesprocessor/go.mod [hunk 10]
```diff
@@ -1,6 +1,6 @@
 module github.com/DataDog/datadog-agent/comp/otelcol/otlp/components/processor/infraattributesprocessor
 
-go 1.25.0
+go 1.24.0
```
Comment (IbraheemA) [inline comp/otelcol/otlp/components/processor/infraattributesprocessor/go.mod:3]: curious why this downgrade?

File: pkg/trace/api/otlp.go (hunk 7 above)
Comment (IbraheemA) [reply pkg/trace/api/otlp.go:303]: Agreed with Nathan here, customers having `k8s.pod.uid` be used as container ID is probably wrong but it does happen in some existing setups, and differing logic between traces/trace metrics would be confusing.

The `OTLPTracesToConcentratorInputsWithObfuscation` method which calls GetContainerID does have access to Agent flags, you could wrap the call there with your flag as a guard.

File: pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go (hunk 6 above)
Comment (IbraheemA) [inline pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go]: nit: Could we return a pcommon.Map here instead of a Resource? (improve isolation)

File: pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go [hunk 11]
```diff
@@ -296,6 +304,78 @@ func ContainerTagsFromResourceAttributes(attrs pcommon.Map) map[string]string {
...
+		if _, found := containerDDTags[key]; found {
+			tagSource = containerTagSourceDD
+			mappedKey = key
+		}
+
+		if tagSource == containerTagSourceNone {
+			return false
+		}
+
+		if tagSource > tagSources[mappedKey] {
```
Comment (IbraheemA) [inline pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go:367]: clever solution!

File: pkg/trace/api/otlp.go [hunk 12]
```diff
@@ -340,45 +301,97 @@ func (o *OTLPReceiver) receiveResourceSpansV2(ctx context.Context, rspans ptrace
...
+			ddspan := transform.OtelSpanToDDSpan(otelspan, otelres, libspans.Scope(), o.conf)
+
+			if p, ok := ddspan.Metrics["_sampling_priority_v1"]; ok {
+				priorityByID[traceID] = sampler.SamplingPriority(p)
+			}
+			tracesByID[traceID] = append(tracesByID[traceID], ddspan)
+		}
+	}
+
+	lang := traceutilotel.GetOTelAttrVal(resourceAttributes, true, string(semconv.TelemetrySDKLanguageKey))
```
Comment (IbraheemA) [inline pkg/trace/api/otlp.go]: nit: should `resourceAttributes` here and below be `otelres.Attributes()`? Shouldn't affect functionality since the keys referenced here are not container tags, but slightly clearer IMO

**Review by IbraheemA — APPROVED** [review]
Approved with nits 🤏

File: pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go (hunk 4 above)
Comment (dineshg13) [reply pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go]: There are two Trace agent components . For OTLP , core agent sends data to Trace Agent. But for DDOT, we control the initialization , we set it here https://github.com/DataDog/datadog-agent/blob/617cb8fc7316bc9cb099d04e1033474b9ba9bf30/cmd/otel-agent/subcommands/run/command.go#L239 . https://github.com/DataDog/datadog-agent/blob/617cb8fc7316bc9cb099d04e1033474b9ba9bf30/comp/trace/config/setup.go#L126 . containerTags function in set as nil OSS collector , we can do same for DDOT.

File: comp/otelcol/otlp/components/processor/infraattributesprocessor/go.mod (hunk 10 above)
Comment (jade-guiton-dd) [reply comp/otelcol/otlp/components/processor/infraattributesprocessor/go.mod:3]: I don't fully understand the reasoning, but I had a series of CI failures that forced it. It looks like making the test in `pkg/trace/otel` depend on `infraattributesprocessor` meant that `infraattributesprocessor` is now considered "used_by_otel" (see the change in `modules.yml`), which means it has to align to the Go version used by the upstream Collector, which right now is 1.24.

File: pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go (hunk 6 above)
Comment (jade-guiton-dd) [reply pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go]: That's what I had at first, but I couldn't find a nice way to do it.
1. I could copy the resource in `receiveResourceSpansV2`, then modify the attributes map in-place, but Andrey argued that the function modifying an input argument in-place was confusing;
2. I could copy the resource in `receiveResourceSpansV2`, then return a new copy of the attributes map from `RemoveContainerTagsFromResource`, but this means we end up copying the map twice;
3. I could return a modified copy of the attributes map from `RemoveContainerTagsFromResource`, then build a new resource with it and manually copy the other fields (dropped attributes count and entity refs).

I figured that taking in the whole resource was fine, since at some point in the future we might want to find container tags by looking at `container` entities in the `Resource`, instead of just the attributes? But maybe that was a bit too forward-thinking. Not sure what kind of "isolation" you're thinking of, but I could go back to solution 1 if that sounds good to you.

File: pkg/trace/api/otlp.go (hunk 12 above)
Comment (jade-guiton-dd) [reply pkg/trace/api/otlp.go]: I wanted to avoid modifying unrelated existing code besides just moving it, but I agree. I can either make sure I reassign `resourceAttributes` and not just `otelres`, or remove `resourceAttributes` and use `otelres.Attributes()` everywhere, at a very slight performance penalty. What do you think?

File: pkg/trace/api/otlp.go (hunk 7 above)
Comment (jade-guiton-dd) [reply pkg/trace/api/otlp.go:303]: Makes sense. If this is the only place where `GetOTelContainerID` is used, I can modify the "canonical" version of the function to no longer use `k8s.pod.uid`, and extract the fallback into `OTLPTracesToConcentratorInputsWithObfuscation` when the flag is disabled, if that makes sense to you.

File: pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go (hunk 4 above)
Comment (jade-guiton-dd) [reply pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go]: It sounds like you're assuming that disabling the Tagger is all we need to fix the issue, but that's not the case. Disabling the tagger means we will get almost no container tags, because the tags added by the Infra Attributes Processor follow Datadog conventions and are only converted into span attributes. The problem that `RemoveContainerTagsFromResource` fixes is that 1. it recognizes Datadog-convention attributes as container tags, and 2. it removes them from the resource so they're not turned into span attributes later. This is necessary, even for DDOT.

But you are correct that once this new feature flag becomes the default, we can switch to a nop `ContainerTags` function in DDOT's instance of the trace agent. I'm not sure if we can gate the initialization with a feature flag in this PR already?

File: pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go (hunk 6 above)
Comment (IbraheemA) [reply pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go]: Hmm, no I agree that solution 1 would be less readable. Good analysis, let's keep it as-is.

File: pkg/trace/api/otlp.go (hunk 12 above)
Comment (IbraheemA) [reply pkg/trace/api/otlp.go]: Since we're mutating `otelres` throughout the function, it would be most readable to just use otelres.Attributes() everywhere IMO.

File: pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go [hunk 13]
```diff
@@ -296,6 +304,76 @@ func ContainerTagsFromResourceAttributes(attrs pcommon.Map) map[string]string {
...
+// 3. Datadog semantic conventions (pre-mapped tags, usually from the infraattributes processor).
+//
+// Only string-type resource attributes will be extracted as container tags.
+// In the case of duplicates between the three sources, OTel conventions take priority over custom tags,
+// which take priority over pre-mapped tags.
+//
+// This function also modifies the input map, removing the attributes matching sources 2 and 3.
+// These filtered resource attributes should be used when converting spans to Datadog format,
+// to avoid attributes being duplicated as both span attributes and container tags.
+func RemoveContainerTagsFromAttributes(attrs pcommon.Map) map[string]string {
```
Comment (IbraheemA) [inline pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go]: super nit: maybe name this "Extract/Consume" rather than remove? With the current name it's unclear that it also returns them IMO

Comment (jade-guiton-dd) [reply pkg/opentelemetry-mapping-go/otlp/attributes/attributes.go]: Agreed, renamed it to "Consume"

**Review by IbraheemA — APPROVED** [review]

**Review by dineshg13 — APPROVED** [review]

File: releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    Added the ``enable_otlp_container_tags_v2`` feature flag, which reduces duplication of container tags in OTLP traces.
```
Comment (songy23) [inline releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml]: You only mentioned OTLP traces - Do the duplicated container tags apply to APM stats as well? Looks like you made the same changes in APM stats

File: pkg/trace/stats/otel.go [hunk 15]
```diff
@@ -64,7 +64,12 @@ func OTLPTracesToConcentratorInputsWithObfuscation(
 		env := transform.GetOTelEnv(otelspan, otelres, conf.OTLPReceiver.IgnoreMissingDatadogFields)
 		hostname := transform.GetOTelHostname(otelspan, otelres, conf.OTLPReceiver.AttributesTranslator, conf.Hostname, conf.OTLPReceiver.IgnoreMissingDatadogFields)
 		version := transform.GetOTelVersion(otelspan, otelres, conf.OTLPReceiver.IgnoreMissingDatadogFields)
-		cid := transform.GetOTelContainerID(otelspan, otelres, conf.OTLPReceiver.IgnoreMissingDatadogFields)
+		var cid string
+		if conf.HasFeature("enable_otlp_container_tags_v2") {
+			cid = transform.GetOTelContainerID(otelspan, otelres, conf.OTLPReceiver.IgnoreMissingDatadogFields)
+		} else {
+			cid = transform.GetOTelContainerOrPodID(otelspan, otelres, conf.OTLPReceiver.IgnoreMissingDatadogFields)
+		}
```
Comment (songy23) [inline pkg/trace/stats/otel.go:72]: Would be good to have a test on this

**Review by songy23 — APPROVED** [review]
Minor questions, otherwise LGTM

File: releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml (hunk 14 above)
Comment (jade-guiton-dd) [reply releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml]: The tagger and our `ContainerTagsFromResourceAttributes` function do not return duplicate tags: the problem is just that we have one set of container tags (the ones from the Infra Attributes Processor) that end up as span attributes on each span, and another set (the ones from the trace agent) which end up as container tags, stored under `_dd.tags.container` at the chunk level.

So because APM stats only have one location to store container tags in, I don't think they suffer from the same problem, although I haven't checked in-depth.

The only change I made to APM stats was changing how the container ID is extracted: with the feature flag on, we stop looking for the container ID in `k8s.pod.uid`, which has never made sense semantically. It's a completely independent change that has nothing to do with tag duplication; it's just that it always bugged me, and a breaking change on container tag handling seemed like a good opportunity to fix it.

**Review by ajgajg1134 — APPROVED** [review]

File: pkg/trace/stats/otel.go (hunk 15 above)
Comment (RiantZ) [reply pkg/trace/stats/otel.go:72]: Are you talking about unit-test, E2E or manual?
I did few tests using Kind cluster with different options on/off.

File: releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml (hunk 14 above)
Comment (nathan-b) [inline releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml]: Would it make sense to have documentation somewhere (or just included in the RNs) about how a customer should know if this flag should be enabled? If I were a customer and I read this line, I'd wonder if my env was duplicating data and would have no way of finding out other than contacting support.

**Review by nathan-b — APPROVED** [review]
One non-blocking question.

File: releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml (hunk 14 above)
Comment (jade-guiton-dd) [reply releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml]: Good point, I'll elaborate in the release note.

Comment (jade-guiton-dd) [reply releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml]: I tried adding more information, tell me what you think.

File: pkg/trace/stats/otel.go (hunk 15 above)
Comment (songy23) [reply pkg/trace/stats/otel.go:72]: Unit tests with this FF on and off should be good. You can add it to https://github.com/DataDog/datadog-agent/blob/main/pkg/trace/otel/stats/otel_util_test.go

File: releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml (hunk 14 above)
Comment (nathan-b) [reply releasenotes/notes/otlp-container-tags-v2-0d02960ec7d67c4e.yaml]: I think this RN does a good job of explaining the tradeoffs. Thanks for taking a second pass!

File: pkg/trace/stats/otel.go (hunk 15 above)
Comment (jade-guiton-dd) [reply pkg/trace/stats/otel.go:72]: Not sure how much value this test has, but I added one

**Review by songy23 — APPROVED** [review]
